src/vft.py

```diff
@@ -91,9 +91,18 @@ def generate_graycode(width, height, vft_id, tag_border_size, value, debug):
 
 
 def analyze_graycode(img, luma_threshold, lock_layout=False, debug=0):
-    bit_stream, vft_id = analyze(img, luma_threshold, lock_layout, debug)
+    global vft_id
+    bit_stream, vft_id = analyze(
+        img, luma_threshold, lock_layout=lock_layout, debug=debug
+    )
     # convert gray code in bit_stream to a number
-    num_read = gray_bitstream_to_num(bit_stream)
+    try:
+        num_read = gray_bitstream_to_num(bit_stream)
+    except (InvalidGrayCode, SingleGraycodeBitError, NoValidTag) as e:
+        print(f"{type(e).__name__}: {e}")
+        if lock_layout and vft_id:
+            # Rerun without locking the layout
+            return analyze_graycode(img, luma_threshold, lock_layout=False, debug=debug)
```

Comment:
Won't this run infinitely?

Reply:
No. It only runs if lock_layout is enabled which in this single recursion it is not.
The idea here is that when the parsing fails it tries again without the locked layout setting.